server/src/test/java/org/apache/calcite/avatica/ha/ConnectionPropertiesHATest.java

```diff
@@ -0,0 +1,311 @@
+
+    String url = SERVERS.getJdbcUrl(START_PORT, Driver.Serialization.PROTOBUF);
+    AvaticaConnection connection = (AvaticaConnection) DriverManager.getConnection(url,
+        properties);
+    String uri = getConnectionURI(connection);
+    Assert.assertEquals(urls[0], uri);
+  }
+
+  @Test
+  public void testConnectionPropertiesHAHttpConnectionTimeout5Sec() throws Exception {
+    //Skip the test for Windows.
+    Assume.assumeFalse(OS_NAME_LOWERCASE.startsWith(WINDOWS_OS_PREFIX));
+    Properties properties = new Properties();
+
+    properties.put(BuiltInConnectionProperty.USE_CLIENT_SIDE_LB.name(), "true");
+    properties.put(BuiltInConnectionProperty.HTTP_CONNECTION_TIMEOUT.name(), "5000");
+    properties.put(BuiltInConnectionProperty.LB_CONNECTION_FAILOVER_RETRIES.name(), "0");
+    //240.0.0.1 is special URL which should result in connection timeout.
+    properties.put(BuiltInConnectionProperty.LB_URLS.name(), "http://240.0.0.1:" + 9000);
+    String url = SERVERS.getJdbcUrl(START_PORT, Driver.Serialization.PROTOBUF);
+    long startTime = System.currentTimeMillis();
+    try {
+      DriverManager.getConnection(url, properties);
+    } catch (RuntimeException re) {
+      long endTime = System.currentTimeMillis();
+      long elapsedTime = endTime - startTime;
+      Assert.assertTrue(elapsedTime < Timeout.ofMinutes(3).toMilliseconds());
+      Assert.assertTrue(elapsedTime >= 5000);
+      Assert.assertTrue(re.getCause() instanceof ConnectTimeoutException);
+    }
+  }
+
+  @Test
+  public void testConnectionPropertiesCreateStatementAfterDisconnect() throws Exception {
+
+    //Start a new server at port 100 port from the startport
+    int test_server_port = START_PORT + 100;
+    HttpServer avaticaServer = Main.start(SERVER_ARGS, test_server_port,
+        AvaticaProtobufHandler::new);
+
+    Properties properties = new Properties();
+    properties.put(BuiltInConnectionProperty.USE_CLIENT_SIDE_LB.name(), "true");
+    properties.put(BuiltInConnectionProperty.LB_CONNECTION_FAILOVER_RETRIES.name(), "2");
+    properties.put(BuiltInConnectionProperty.LB_CONNECTION_FAILOVER_SLEEP_TIME.name(), "100");
+    properties.put(BuiltInConnectionProperty.HTTP_CONNECTION_TIMEOUT.name(), "300");
+    properties.put(BuiltInConnectionProperty.LB_STRATEGY.name(),
+        RoundRobinLBStrategy.class.getName());
+
+    StringBuilder sb = new StringBuilder();
+    //First URL will be server we started in this test
+    sb.append("http://localhost:").append(test_server_port).append(",");
+    for (int i = 0; i < 1000; i++) {
+      sb.append("http://localhost:").append(START_PORT + i).append(",");
+    }
+    properties.put(BuiltInConnectionProperty.LB_URLS.name(), sb.substring(0, sb.length() - 1));
+
+    //Create a connection
+    String url = SERVERS.getJdbcUrl(test_server_port, Driver.Serialization.PROTOBUF);
+    AvaticaConnection conn = (AvaticaConnection) DriverManager.getConnection(url, properties);
+
```

Comment:
Better have a test here first here, to make sure that the connection works before we shut the server down.
(Just create a connection, and select anything)

Reply:
Actually, to exercise the most code paths, and make sure we do not transparently reconnect:
Create connection1
Create statement1
Execute select
Kill server
Try to execute a query on statement1 (expect to fail)
Try to create statement2 on connection1(expect to fail)

Reply:
Done